cirq/protocols/qid_shape_protocol.py

```diff
@@ -160,14 +162,27 @@ def num_qubits(
             method (or they returned NotImplemented) and also no default value
             was specified.
     """
+    # Helper to call a optional getter function when val is either a Class or an instance
+    def call_getter(opt_getter_func):
+        if opt_getter_func is None:
+            return NotImplemented
+
+        if not isclass(val):
+            return opt_getter_func()
+
+        try:
+            return opt_getter_func(None)
+        except AttributeError:
```

Comment:
I'm a little leery about using a try-except block here since `AttributeError` could be raised for reasons other than the one we have in mind here. It'd be safer to directly check the presence of the attribute using `hasattr`.

(This is similar to the EAFP idiom for checking key presence in a dict, so it's certainly idiomatic. It seems unnecessarily risky, though.)